testnet/stacks-node/src/run_loop/puppet.rs

```diff
@@ -0,0 +1,176 @@
+                                        }
+                                        Ok(Response::new(StatusCode::Ok))
+                                    }
+                                    (Method::Put, "/until") => {
+                                        let body = req.body_string().await;
+                                        match body {
+                                            Ok(x) => {
+                                                let v = x.parse::<u64>().unwrap_or(0);
+                                                if v > 0 {
+                                                    println!("Setting target block to {}", v);
+                                                    io::stdout().flush().unwrap();
+                                                    let mut puppet_control =
+                                                        puppet_control.lock().unwrap();
+                                                    puppet_control.target_block = v;
+                                                }
+                                            }
+                                            _ => (),
+                                        }
+                                        Ok(Response::new(StatusCode::Ok))
+                                    }
+                                    _ => {
+                                        let mut rs = Response::new(StatusCode::BadRequest);
+                                        rs.set_body(format!(
+                                            "[{}] {}",
+                                            req.method(),
+                                            req.url().path()
+                                        ));
+                                        Ok(rs)
+                                    }
+                                }
+                            })
+                            .await
+                            .unwrap_or(())
+                        }
+                        Ok(())
+                    })
+                })
+                .unwrap(),
+        )
+    }
+
+    pub fn block_on(&self, chain_tip: &ChainTip) {
+        if self.join_handle.is_none() {
+            return;
+        }
+        info!(
+            "Waiting on block height {}",
+            chain_tip.metadata.stacks_block_height
+        );
+        let mut puppet_control = self.inner.lock().unwrap();
+        puppet_control.current_block = chain_tip.metadata.stacks_block_height;
+        drop(puppet_control);
+        loop {
+            let mut puppet_control = self.inner.lock().unwrap();
+            if puppet_control.target_block > puppet_control.current_block
+                || puppet_control.next_block_time.le(&SystemTime::now())
+            {
+                puppet_control.next_block_time =
+                    SystemTime::now().add(puppet_control.block_interval);
+                break;
+            }
+            drop(puppet_control);
```

Comment:
I'm not sure this `drop()` is necessary? Won't `puppet_control` go out of scope at the end of the loop?

Reply:
This is necessary because of the next line `sleep`. Refactored with the `with_lock` function mentioned above.